### ISSUE TITLE
Fixing `'shell'` on Windows, properly

### ISSUE THREAD

**Issue by LunarLambda** [neovim/neovim#28384]
### Problem

See https://github.com/neovim/neovim/issues/16957 for some of the existing issues with using a Unix shell in Neovim on Windows. 

I've dug into it now, seeing as no workaround really managed to fully take, and here's what I was able to find.

The root of evil is the fact that `CreateProcessA` takes some long string as the command line. The Microsoft C runtime takes that string, runs it through splitting logic described [here](https://learn.microsoft.com/en-us/cpp/c-language/parsing-c-command-line-arguments?view=msvc-170), then passes the processed result to `main`. The result of that processing is what `'shell'` ultimately receives.

Vim on the other hand, wants to be able to think about running programs in terms of regular C `argv` arrays.

Assuming `shell=bash`, `shellcmdflag=-c` and `shellxquote="`, running `!ls -l`, produces the argument vector `[bash, -c, "ls -l"]` (note the quotes around `ls -l`). The resulting argument vector is then [passed to libuv](https://github.com/neovim/neovim/blob/master/src/nvim/event/libuv_process.c#L22), which then [reconstitutes the entire thing back to a single string](https://github.com/libuv/libuv/blob/v1.x/src/win/process.c#L521) so it can pass it to `CreateProcessA`.

This is where the issue arises. libuv expects an *unquoted* argument vector (`[bash, -c, ls -l]`, no quotes), as you would pass to `execve` on Unix, but gets `[bash, -c, "ls -l"]` instead. This causes it to *quote the quoted string* on Windows, which ultimately causes bash to receive the command `"ls -l"` (one shell word, with quotes) instead of `ls -l` (two shell words, no quotes).

So, basically, there are two places in the chain that are trying to make the argument vector "windows-friendly": Vim's `shell_build_argv`, and libuv's `make_program_args`. Only one of the two should be done. To only use vim's behaviour, it is probably sufficient to change [libuv_process_spawn](https://github.com/neovim/neovim/blob/master/src/nvim/event/libuv_process.c#L32) to always pass `UV_PROCESS_WINDOWS_VERBATIM_ARGUMENTS`. Since `cmd.exe` already has to use this behaviour since its argument parsing does *not* follow that of the C runtime (i.e. all other programs), this seems like it will cause the least amount of breakage.

To only follow libuv's behaviour (an arguably more sound strategy but that may not work with cmd.exe), we'd simply need to skip `shell_xescape_xquote` (or set both `shellxescape` and `shellxquote` to the empty string?).

Whichever the case may be, it's pretty clear that Neovim's current behaviour on Windows is incorrect. Double quoting like this shouldn't occur, and having two completely different implementations both doing argv-sanitation depending on context is bound to lead to issues. (As it did very recently for Rust)

There's also some other issues. For example, Neovim tries to pass `2>&1 | tee <tempfile>` to the shell. However, the Windows-style backslash-using tempfile path will cause the shell to misinterpret something like `C:\somedir\tmp\` to contain a `\t` (tab) (at least in zsh) and cause it to spit out the file in the wrong place, causing errors in `:make`, `:grep`, etc. Easiest solution would be to change the path to use `/` as the separator (possibly in accordance with `'shellslash'`, which doesn't apply to the temporary file arguments currently) (though this will also not work for UNC paths), or to use something like `cygpath` if available (most unix shells on Windows are cygwin/msys2 based).

### Steps to reproduce

nvim --clean
:set shell=C:/path/to/bash.exe shellcmdflag=-c
:!echo test <- fails
:set shellxquote=
:!echo test <- works
(and many other such cases)

### Expected behavior

We should figure out a way to fix this properly, once and for all.

### Neovim version (nvim -v)

master

### Vim (not Nvim) behaves the same?

Have not verified, but it seems to be Neovim specific

### Operating system/version

Windows 10/11

### Terminal name/version

wezterm, but doesn't matter (not TUI related)

### $TERM environment variable

xterm-256color

### Installation

GitHub nightly releases

**Comment by clason** [comment]
For the record, being more restrictive (and explicit) about which "shells" we support on Windows is an option here (though maybe not for 0.10).

**Comment by LunarLambda** [comment]
In the sense of saying "we only support cmd.exe and powershell", or "we eventually stop supporting cmd.exe"?

I use Neovim on Windows exclusively through an MSYS2 environment, and while I understand it's not an officially supported use case/"platform", I would really like for unix shells to be supported and working. Lots of people use Git Bash on Windows for example.

I think, generally speaking:

For non-cmd.exe, we should completely ignore shellquote/shellxquote/shellxescape/etc, not do any quoting/escaping of our own, and instead let libuv handle it for us. This will work correctly with any program that uses the regular C runtime argument handling (including, I believe, powershell/pwsh).

For cmd.exe, we could continue using the existing code paths and tell libuv to pass the arguments verbatim as we do now. libuv does not handle cmd.exe specially in any way, so this seems reasonable if it's important that the edge cases are handled proper.

I could imagine a future where Neovim simply doesn't support cmd.exe as a shell anymore, and defaults to powershell. I can't think of anything that would definitely break, since users are already free to change the shell settings away from cmd.exe, so nothing inside of Neovim should really have a hard dependency on it, but I still imagine it may cause *some* churn.

Either way, that decision should not stand in the way of fixing the existing use cases today.

**Comment by clason** [comment]
> I could imagine a future where Neovim simply doesn't support cmd.exe as a shell anymore, and defaults to powershell.

That's what I was thinking, yes, _if_ it allows things to work better on other platforms (including win shells).

**Comment by justinmk** [comment]
Nice analysis!

> So, basically, there are two places in the chain that are trying to make the argument vector "windows-friendly": Vim's `shell_build_argv`, and libuv's `make_program_args`. Only one of the two should be done. To only use vim's behaviour, it is probably sufficient to change [libuv_process_spawn](https://github.com/neovim/neovim/blob/master/src/nvim/event/libuv_process.c#L32) to always pass `UV_PROCESS_WINDOWS_VERBATIM_ARGUMENTS`. Since `cmd.exe` already has to use this behaviour since its argument parsing does _not_ follow that of the C runtime (i.e. all other programs), this seems like it will cause the least amount of breakage

Based on the git logs, I don't see a rationale for why 
 https://github.com/neovim/neovim/commit/f3cc843755a6d638ada77dc31721aa53b3ff2364 special-cased "cmd.exe", I assume it was just conservative.



> For non-cmd.exe, we should completely ignore shellquote/shellxquote/shellxescape/etc, not do any quoting/escaping of our own, and instead let libuv handle it for us.

Sure, let's try it. Existing tests are here: https://github.com/neovim/neovim/blob/8e5c48b08dad54706500e353c58ffb91f2684dd3/test/functional/vimscript/system_spec.lua#L130

>  "we eventually stop supporting cmd.exe"

Seems like a separate topic. Based on this thread, I don't see a reason to drop support for a particular shell?

**Comment by LunarLambda** [comment]
One thing I'd definitely like to fix is this

```vim
:make
:!cargo  2>&1| tee F:\msys2\tmp\nvim.0\cn0m97\2
cargo  2>&1| tee F:\msys2       mp
vim.0error
```
The weird printing seems to be zsh specific, but even with bash it still doesn't write the correct file, so quickfix is broken.

I think the temporary file path should be obeying `'shellslash'`:

> When set, a forward slash is used when expanding file names.  This is
> useful when a Unix-like shell is used instead of cmd.exe.  Backward
> slashes can still be typed, but they are changed to forward slashes by

I'm not sure how involved of a fix that is, but seeing as Neovim doesn't currently use UNC paths, I don't see a big issue.

**Comment by clason** [comment]
> I think the temporary file path should be obeying 'shellslash':

Or `'shellslash'` should be removed and forward slashes be used unconditionally. The point here is that we are fine making breaking changes for legacy systems that are no longer officially supported, _if_ that helps fixing bugs on supported systems.

**Comment by clason** [comment]
(Again, this is just opening a door to more options for a solution to these issues, not a push to do that.)

**Comment by dundargoc** [comment]
> Note we claim support for Windows >= 8, but that, too, can be revisited.

I will one-up you and claim we already don't support anything older than Windows 10 "Version 1809" as the `:terminal` feature does not work for those systems. It depends on how one interprets what "support" means, but I personally interpret it to mean ALL editor features are supported, rather than that some parts happen to work.

One might argue that this quasi-bump in minimum version was accidental and that it should therefore count more as a bug/accident than intentional. I personally think this is a bit weak stance as documenting this problem is already an admission from our side that we don't see this being resolved soon.

I propose we simply bite the bullet and officially only support Windows 10 "Version 1809" to simplify the question on which windows version we support (which is brought up surprisingly frequently).

**Comment by clason** [comment]
>  It depends on how one interprets what "support" means, but I personally interpret it to mean ALL editor features are supported, rather than that some parts happen to work.

Yes. The current wording implies that ":terminal" is not covered by "support". It's a slippery slope I would like to get off of. It strikes me as weird that we support significantly more versions than our core dependencies (libuv; would have to check tree-sitter).

**Comment by LunarLambda** [comment]
Linking https://github.com/neovim/neovim/pull/28395 since the topic of slash handling got brought up here.

Treating this as a sort of "tracking issue" while the details of what we want to change/fix aren't fully decided yet.

If/when that PR lands, I would try and take stock of what works/doesn't on Windows with various shells and try to implement the "let's have libuv alone handle non-cmd shells" strategy, unless there are strong objections or other things to try first.

**Comment by ykhan21** [comment]
If I run `:!ls` in neovim (installed with scoop) in bash (from Git for Windows, installed with scoop), I get:
```
:!ls
/usr/bin/bash: /s: No such file or directory

shell returned 127
```

@LunarLambda, is this the same error in this issue?

The error message above reminds me of a cmd-bash thing where in bash, to run cmd commands, you need to escape the forward slashes that prefix flags for cmd commands. For example:
In cmd you would run: `tasklist /fo csv`
In bash you could run: `taskkill //fo csv`

In bash, if you run `tasklist /fo csv`, you get the error:
```sh
$ tasklist /fo csv
ERROR: Invalid argument/option - 'C:/Users/USERNAME/scoop/apps/git/2.45.1/fo'.
Type "TASKLIST /?" for usage.
```

**Comment by LunarLambda** [comment]
It is the same error. For a workaround, you can set the following: `:set shellcmdflag=-c shellxquote= shellxescape=`

This is enough to get git bash and other posix-ish shells working on Windows. It might not be a perfect solution, because as described in this issue, Neovim is currently doing things in a very roundabout way.

**Comment by ykhan21** [comment]
This fixes `:!...` commands, but not `:'<,'>!...` commands.
For example, in vim, you could highlight some text with visual mode and run `:!wc` (which actually runs `:'<,'>!wc`).

**Comment by LunarLambda** [comment]
Quite possible, input redirections done by neovim when calling the shell are often broken due to shells intepreting the \\ path separator incorrectly.

**Comment by ykhan21** [comment]
Yes, it seems like it. This is the error:
```
:'<,'>!wc
shell returned 1

E485: Can't read file C:\Users\USERNAME\AppData\Local\Temp\nvim.0\olxDWI\1
```

**Comment by LunarLambda** [comment]
Yes. This is already explained in some of the initial comments to this issue. There is a linked PR that would fix this by making Neovim use / always on Windows, but it is currently stalled.

**Comment by xudyang1** [comment]
@LunarLambda  @ykhan21 

For people who want to use git-bash for `:!` or `:ter` on Windows, the following config works.

For zsh or other unix shell, setting `opts.shellslash = true` might also help.

 ```lua
if vim.fn.has("win32") == 1 or vim.fn.has("win64") == 1 then
  local opt = vim.opt
  -- use git-bash for `:!`, `:ter`
  -- windows only, convert \ to / when expanding file names
  opt.shellslash = true
  -- defaults to cmd.exe
  opt.shell = vim.fs.normalize("~/scoop/apps/git/current/bin/bash.exe")
  -- opt.shell = "bash.exe" -- WARN: sometimes mistakenly points to wsl bash

  -- defaults to "/s /c" for cmd.exe
  opt.shellcmdflag = "-c"
  -- default value works
  opt.shellpipe = "2>&1| tee"
  -- Windows: may default to "\"" when 'shell' contains 'sh'
  opt.shellquote = ""
  opt.shellxquote = ""
  -- defaults to ">", 'bash' uses:
  opt.shellredir = ">%s 2>&1"
  -- defualt value works
  opt.shellxescape = ""
end
```
